pypesto/engine/mpi_pool.py

```diff
@@ -0,0 +1,61 @@
+from multiprocessing import Pool
+from mpi4py.futures import MPIPoolExecutor
+from mpi4py import MPI
+import cloudpickle as pickle
+import logging
+from typing import List
+
+from .base import Engine
+from .task import Task
+
+
+logger = logging.getLogger(__name__)
+
+
+def work_per_node(chunk):
+    with Pool() as pool:
+        results = pool.map(work_per_core, chunk)
+    return results
+
+
+def work_per_core(pickled_task):
+    """Unpickle and execute task."""
+    task = pickle.loads(pickled_task)
+    return task.execute()
+
+
+class MPIPoolEngine(Engine):
+    """
+    Parallelize the task execution using multiprocessing.
+
+    Parameters
+    ----------
+    chunksize:
+        The chunksize the MPIPoolExecutor uses.
+        Default is two, but should always be set to
+        the number of cores used in each node.
+    """
+
+    def __init__(self, chunksize: int = 2):
+        super().__init__()
+        self.chunksize: int = chunksize
+
+    def execute(self, tasks: List[Task]):
+        """Pickle tasks and distribute work over nodes."""
+
+        pickled_tasks = [pickle.dumps(task) for task in tasks]
+        # put the tasks in batchsizes of the number of cores used
+        chunks = [pickled_tasks[x:x+self.chunksize]
+                  for x in range(0, len(pickled_tasks), self.chunksize)]
+
+        n_procs = MPI.COMM_WORLD.Get_size()   # Size of communicator
+        logger.info(f"Performing parallel task execution on {n_procs} "
+                    f"nodes with chunksize of {self.chunksize}.")
```

Comment:
Communicator size is not necessarily the number of nodes.

Reply:
so i do think it is, at least when i tested, if i allocated e.g. 5 nodes in my slurm file, the communicator size was also 5. Not sure whether this is exactly want you meant though

Reply:
It depends on how you invoke mpiexec. As [mentioned](https://github.com/ICB-DCM/pyPESTO/pull/542#issuecomment-746891307), I don't see any obvious problem with having one MPI rank per core instead of per node. There slightly more expensive communication should be overcompensated by not wasting one full node.

Reply:
might be because i invoked mpiexec with `-n [Number]`that it always only ran one process per node.